lib/ReorderBasicBlocksPass.cpp

```diff
@@ -41,21 +54,94 @@ namespace clspv {
 FunctionPass *createReorderBasicBlocksPass() {
   return new ReorderBasicBlocksPass();
 }
+} // namespace clspv
+
+void ReorderBasicBlocksPass::StructuredOrder(BasicBlock *block,
+                                             const LoopInfo &LI,
+                                             std::deque<BasicBlock *> *order,
+                                             DenseSet<BasicBlock *> *visited) {
+  if (!visited->insert(block).second)
+    return;
+
+  // Identify the merge and continue blocks for special treatment.
+  const auto *terminator = dyn_cast<TerminatorInst>(block->getTerminator());
+  BasicBlock *continue_block = nullptr;
+  BasicBlock *merge_block = nullptr;
+  if (LI.isLoopHeader(block)) {
+    Loop *loop = LI.getLoopFor(block);
+    merge_block = loop->getExitBlock();
+
+    if (loop->isLoopLatch(block)) {
```

Comment:
Needs some comments about what's going on.

My understanding is:   the "then" case occurs when the block (which is a loop header) is its own latch.  So that's the  case of a single-block loop.  Please add a comment to that effect.